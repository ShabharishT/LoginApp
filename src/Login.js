import React, {Component} from 'react';
import FormErrors from './FormErrors';

/*
 * Login component to validate user inputs,
 * authenticate and route to Home page.
 */
export default class Login extends Component {

    constructor(props) {
        super(props);
        this.state = {
            username: '',
            password : '',
            validUserName: false,
            validPassword: false,
            formValid: false,
            formErrors: {username: '', password: ''}
        };
        this.handleLogin = this.handleLogin.bind(this);
    }

    handleUserInput(e) {
        const name = e.target.name;
        const value = e.target.value;
        this.setState({[name]: value}, () => {this.validateField(name, value)});
    }

    validateField(fieldName, value) {
        let fieldValidationErrors = this.state.formErrors;
        let validUserName = this.state.validUserName;
        let validPassword = this.state.validPassword;

        switch(fieldName) {
            case 'username':
                // Using a valid regex pattern for email pattern
                validUserName = value.match(/^([\w.%+-]+)@([\w-]+\.)+([\w]{2,})$/i);
                fieldValidationErrors.username = validUserName ? '' : ' is invalid';
                break;

            case 'password':
                validPassword = value.length >= 6;
                fieldValidationErrors.password = validPassword ? '': ' is too short';
                break;

            default:
                break;
        }
        this.setState({formErrors: fieldValidationErrors,
            validUserName: validUserName,
            validPassword: validPassword
        }, this.validateForm);
    }

    validateForm() {
        this.setState({formValid: this.state.validUserName && this.state.validPassword});
    }


    handleLogin() {
        this.props.history.push({
            pathname: '/home',
            state: {userName: this.state.username}
        });
    };

    render() {
        const {formErrors, formValid} = this.state;
        return (
            <form className="demoForm">

                <h4>Enter your credentials</h4>

                <div className="panel panel-default">
                    <FormErrors formErrors={formErrors} />
                </div>

                <div className="form-group">
                    <div><label htmlFor="email">Username</label></div>
                    <input type="email" className="form-control" name="username"
                           onChange={(event) => this.handleUserInput(event)}/>
                </div>

                <br/>

                <div className="form-group">
                    <div><label htmlFor="password">Password</label></div>
                    <input type="password" className="form-control" name="password"
                           onChange={(event) => this.handleUserInput(event)}/>
                </div>

                <br/>

                <button type="submit" className="btn btn-primary" disabled={!formValid}
                        onClick={this.handleLogin}>Login</button>

            </form>
        );
    }
}