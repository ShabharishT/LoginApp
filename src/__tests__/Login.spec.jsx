import React from 'react';
import renderer from 'react-test-renderer';
import Login from '../Login';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16'

configure({adapter: new Adapter()});

const usernameInput = {
    target: {
        name: 'username',
        value: 'shabt@vmware.com'
    }
};

const passwordInput = {
    target: {
        name: 'password',
        value: '!@ASDA!@qw1'
    }
};

describe('Login Component ', () => {

    test ('Successful render', () => {
        const component = renderer.create(<Login />);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test ('Handle username', () => {
        const wrapper = shallow(<Login />);
        const layout = wrapper.instance();
        expect(layout.handleUserInput(usernameInput)).toMatchSnapshot();
    });

    test ('Handle password', () => {
        const wrapper = shallow(<Login />);
        const layout = wrapper.instance();
        expect(layout.handleUserInput(passwordInput)).toMatchSnapshot();
    });
});