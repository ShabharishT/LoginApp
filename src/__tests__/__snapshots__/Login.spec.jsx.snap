// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component  Handle password 1`] = `undefined`;

exports[`Login Component  Handle username 1`] = `undefined`;

exports[`Login Component  Successful render 1`] = `
<form
  className="demoForm"
>
  <h2>
    Sign In
  </h2>
  <div
    className="panel panel-default"
  >
    <div
      className="formErrors"
    >
      
      
    </div>
  </div>
  <div
    className="form-group"
  >
    <div>
      <label
        htmlFor="email"
      >
        Username
      </label>
    </div>
    <input
      className="form-control"
      name="username"
      onChange={[Function]}
      type="email"
    />
  </div>
  <br />
  <div
    className="form-group"
  >
    <div>
      <label
        htmlFor="password"
      >
        Password
      </label>
    </div>
    <input
      className="form-control"
      name="password"
      onChange={[Function]}
      type="password"
    />
  </div>
  <br />
  <button
    className="btn btn-primary"
    disabled={true}
    onClick={[Function]}
    type="submit"
  >
    Login
  </button>
</form>
`;
